pyxem/generators/variance_generator.py

```diff
@@ -0,0 +1,128 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017-2019 The pyXem developers
+#
+# This file is part of pyXem.
+#
+# pyXem is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# pyXem is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with pyXem.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Variance generators in real and reciprocal space for fluctuation electron
+microscopy.
+
+"""
+
+import numpy as np
+from hyperspy.signals import Signal2D
+from hyperspy.api import stack
+
+from pyxem.signals.diffraction_variance2d import DiffractionVariance2D
+from pyxem.signals.diffraction_variance2d import ImageVariance
+from pyxem.signals import transfer_signal_axes
+from pyxem.signals import transfer_navigation_axes_to_signal_axes
+
+
+class VarianceGenerator():
+    """Generates variance images for a specified signal and set of aperture
+    positions.
+
+    Parameters
+    ----------
+    signal : ElectronDiffraction2D
+        The signal of electron diffraction patterns to be indexed.
+
+    """
+
+    def __init__(self, signal, *args, **kwargs):
+        self.signal = signal
+
+        # add a check for calibration
+
+    def get_diffraction_variance(self, dqe):
+        """Calculates the variance in scattered intensity as a function of
+        scattering vector.
+
+        Parameters
+        ----------
+
+        dqe : float
+            Detective quantum efficiency of the detector for Poisson noise
+            correction.
+
+        Returns
+        -------
+
+        vardps : Signal2D
+            A DiffractionVariance object containing the mean DP, mean
+            squared DP, and variance DP.
+        """
+        dp = self.signal
+        mean_dp = dp.mean((0, 1))
+        meansq_dp = Signal2D(np.square(dp.data.astype(np.uint16))).mean((0, 1))
+        normvar = (meansq_dp.data / np.square(mean_dp.data)) - 1.
+        var_dp = Signal2D(normvar)
```

Comment:
you can delete this line, and replace `var_dp.data` with `normvar` in the next one